core/src/main/java/io/grpc/PickFirstBalancerFactory2.java

```diff
@@ -0,0 +1,155 @@
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
+
+import java.net.SocketAddress;
+import java.util.List;
+
+/**
+ * A {@link LoadBalancer} that provides no load balancing mechanism over the
+ * addresses from the {@link NameResolver}.  The channel's default behavior
+ * (currently pick-first) is used for all addresses found.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/1771")
+public final class PickFirstBalancerFactory2 extends LoadBalancer2.Factory {
+
+  private static final PickFirstBalancerFactory2 instance = new PickFirstBalancerFactory2();
+
+  private PickFirstBalancerFactory2() {
+  }
+
+  public static PickFirstBalancerFactory2 getInstance() {
+    return instance;
+  }
+
+  @Override
+  public LoadBalancer2 newLoadBalancer(LoadBalancer2.Helper helper) {
+    return new PickFirstBalancer(helper);
+  }
+
+  @VisibleForTesting
+  static class PickFirstBalancer extends LoadBalancer2 {
+    private final Helper helper;
+    private Subchannel subchannel;
+
+    public PickFirstBalancer(Helper helper) {
+      this.helper = helper;
+    }
+
+    @Override
+    public void handleResolvedAddresses(List<ResolvedServerInfoGroup> servers,
+        Attributes attributes) {
+      // Flatten servers list received from name resolver into single address group. This means that
+      // as far as load balancer is concerned, there's virtually one single server with multiple
+      // addresses so the connection will be created only for the first address (pick first).
+      EquivalentAddressGroup newEag =
+          flattenResolvedServerInfoGroupsIntoEquivalentAddressGroup(servers);
+      if (subchannel == null || !newEag.equals(subchannel.getAddresses())) {
+        if (subchannel != null) {
```

Comment:
Does this code here mean that we create a new subchannel while we're closing the old one? Just wondered we should keep an old subchannel while we finish creating a new subchannel.